[libunwind][AArch64] Disable ZA before resuming from unwinding (on Linux)

MacDue opened the pull request:
This patch reimplements the SME ABI `__arm_za_disable` routine within libunwind. This routine must be called before resuming from unwinding on AArch64 platforms with SME support.

Before calling the routine, we need to check that SME is available. In this patch, this is implemented for Linux-based platforms by checking HWCAP2. It should be possible to implement this check for other platforms as required.

This patch includes a test for this functionality. This test requires SME, so on platforms without it, it will simply pass.

sdesmalen-arm commented on libunwind/src/Registers.hpp:
```diff
@@ -1908,25 +1917,43 @@ class _LIBUNWIND_HIDDEN Registers_arm64 {
-    uint64_t __sp;    // Stack pointer x31
-    uint64_t __pc;    // Program counter
-    uint64_t __ra_sign_state; // RA sign state register
+    uint64_t __x[29] = {};        // x0-x28
```
Are these NFC changes necessary?

sdesmalen-arm reviewed (commented):
In general I would prefer the approach taken in #160905 because we'd not have to reimplement the SME ABI routine for `__arm_za_disable`, but that would add a dependency on the compiler-rt `builtins` lib which I'm not sure we can add (any input from libunwind maintainers would be helpful here). In the meantime, I'm happy with the approach of reimplementing the function instead.

MacDue committed:
[libunwind] Disable ZA before resuming from unwinding (on Linux)

This patch reimplements the SME ABI `__arm_za_disable` routine within
libunwind. This routine must be called before resuming from unwinding on
AArch64 platforms with SME support.

Before calling the routine, we need to check that SME is available. In
this patch, this is implemented for Linux-based platforms by checking
HWCAP2. It should be possible to implement this check for other
platforms as required.

This patch includes a test for this functionality. This test requires
SME, so on platforms without it, it will simply pass.

MacDue committed:
Add test again (somehow removed?)

sdesmalen-arm commented on libunwind/src/UnwindRegistersSave.S:
```diff
@@ -827,6 +827,54 @@ DEFINE_LIBUNWIND_FUNCTION(__unw_getcontext)
+//
+  .p2align 2
+DEFINE_LIBUNWIND_FUNCTION(__libunwind_Registers_arm64_za_disable)
+  // If TPIDR2_EL0 is null, the subroutine just disables ZA.
```
> IIRC, pacibsp also acts like a bti

I believe the two have different purposes, where `bti` protects the entry point into the function and `pac` authenticates the return address.

> I'm not sure what .variant_pcs would mean/do here

The function preserves more registers than the normal calling convention. If lazy-binding is enabled then the resolver function that finds the address, updates the GOT and calls this function should also preserve the same registers as the function that is being called. Otherwise, the first call to this function does not preserve the same registers.

https://github.com/ARM-software/abi-aa/blob/main/aaelf64/aaelf64.rst#st-other-values
> If a subroutine is called via a symbol reference that is marked with STO_AARCH64_VARIANT_PCS, then code that runs between the calling routine and the called subroutine must preserve the contents of all registers except for IP0, IP1, and the condition code flags [[AAPCS64](https://github.com/ARM-software/abi-aa/releases)]. It must also preserve the processor execution mode, such as PSTATE.SM and PSTATE.ZA. It is not possible to provide a definitive list of which state must be preserved, since the intention is to allow it to grow as the architecture evolves.
> 
> In particular, when a call is made via the PLT entry of a symbol marked with STO_AARCH64_VARIANT_PCS, a dynamic linker cannot assume that the call follows the register usage convention of the base procedure call standard.

MacDue replied:
> I believe the two have different purposes, where `bti` protects the entry point into the function and `pac` authenticates the return address.

I was referring to `pacibsp` acting as an implicit `bti` see:
https://developer.arm.com/documentation/ddi0602/2025-09/Base-Instructions/PACIB--PACIB1716--PACIBSP--PACIBZ--PACIZB--Pointer-Authentication-Code-for-instruction-address--using-key-B-

However, I've also just noticed:
```
#define DEFINE_LIBUNWIND_FUNCTION(name)                                        \
  .globl SYMBOL_NAME(name) SEPARATOR                                           \
  HIDDEN_SYMBOL(SYMBOL_NAME(name)) SEPARATOR                                   \
  SYMBOL_IS_FUNC(SYMBOL_NAME(name)) SEPARATOR                                  \
  PPC64_OPD1                                                                   \
  SYMBOL_NAME(name):                                                           \
  PPC64_OPD2                                                                   \
  AARCH64_BTI
#endif
```

So the `DEFINE_LIBUNWIND_FUNCTION` will also add a `bti`.

---

> The function preserves more registers than the normal calling convention. If lazy-binding is enabled then the resolver function that finds the address, updates the GOT and calls this function should also preserve the same registers as the function that is being called. Otherwise, the first call to this function does not preserve the same registers.

I've added `.variant_pcs` for completeness, though I don't believe lazy binding will be used here (I think this is a local symbol statically linked into libunwind).

MacDue committed:
Add .variant_pcs
